### PR TITLE
[TS] LPS-162706

### PR DESCRIPTION
**Pull request by yuhai**
@dantewang,  `Files.createSymbolicLink `should be a limitation for windows. Please refer to https://bugs.openjdk.org/browse/JDK-8218418. If ci test doesn't run the test in windows, may I ignore it? If we can't ignore it, can we use Assume to skip it?

@vicnate5, will ci run this test under windows?

cc @ericyanLr

### CODE REVIEW

**Comment by yuhai** [comment]
@dantewang, I just push force the changes about variable naming, no logic changes.

File: modules/apps/static/portal-configuration/portal-configuration-persistence-impl/src/main/java/com/liferay/portal/configuration/persistence/internal/ConfigurationPersistenceManager.java [hunk 1]
```diff
@@ -410,14 +407,10 @@ private void _deleteFromDatabase(String pid) throws IOException {
...
-			PropsValues.MODULE_FRAMEWORK_CONFIGS_DIR);
-
-		configsDirPath = configsDirPath.toRealPath();
-
-		Path configFilePath = configsDirPath.resolve(fileName);
+		File configFile = new File(
+			PropsValues.MODULE_FRAMEWORK_CONFIGS_DIR, fileName);
 
-		return configFilePath.toFile();
+		return configFile.getAbsoluteFile();
```
Comment (ericyanLr) [inline modules/apps/static/portal-configuration/portal-configuration-persistence-impl/src/main/java/com/liferay/portal/configuration/persistence/internal/ConfigurationPersistenceManager.java:413]: Hi @yuhai ,

Just wanted to make sure, would this change re-introduce the issue in [LPS-113749](https://issues.liferay.com/browse/LPS-113749)?

The full test would be to combine steps from both [LPS-113749](https://issues.liferay.com/browse/LPS-113749) and [LPS-162706](https://issues.liferay.com/browse/LPS-162706)

Comment (ericyanLr) [reply modules/apps/static/portal-configuration/portal-configuration-persistence-impl/src/main/java/com/liferay/portal/configuration/persistence/internal/ConfigurationPersistenceManager.java:413]: **Additional Info - Debugging**
For my testing, I found it to be important to test during startup. Here are two breakpoints I used to track how felix fileinstall filename is stored:
- [ConfigurationFileInstaller.java#L141-L149](https://github.com/liferay/liferay-portal/blob/7.4.3.42-ga42/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/configuration/ConfigurationFileInstaller.java#L141-L149)
- [ConfigurationPersistenceManager.java#L614-L620](https://github.com/liferay/liferay-portal/blob/7.4.3.42-ga42/modules/apps/static/portal-configuration/portal-configuration-persistence-impl/src/main/java/com/liferay/portal/configuration/persistence/internal/ConfigurationPersistenceManager.java#L614-L620)

**Comment by vicnate5** [comment]
@yuhai CI has no support for Windows and we have no plans to add support.

File: modules/apps/static/portal-configuration/portal-configuration-persistence-impl/src/main/java/com/liferay/portal/configuration/persistence/internal/ConfigurationPersistenceManager.java (hunk 1 above)
Comment (yuhai) [reply modules/apps/static/portal-configuration/portal-configuration-persistence-impl/src/main/java/com/liferay/portal/configuration/persistence/internal/ConfigurationPersistenceManager.java:416]: Hi @ericyanLr,

Thanks for pointing out the issue. The previous fix didn't work for [LPS-113749](https://issues.liferay.com/browse/LPS-113749).  I just add the new last commit to fix the case. Please see https://github.com/liferay-core-infra/liferay-portal/pull/1772/commits/99edf76f10dacd337ea947c4cfc0e6c8ad6d124c

**Comment by yuhai** [comment]
@dantewang, @ericyanLr, 
If one file is symlink and **file.getCanonicalFile()** want to get the real file, instead of getting symlink file. The cases won't work in windows because **file.getCanonicalFile()** will return the symlink file.

For the [LPS-113749](https://issues.liferay.com/browse/LPS-113749), portal can't startup successful in windows because the [code ](https://github.com/yuhai/liferay-portal/blob/LPS-162706-3/modules/core/portal-bootstrap/src/main/java/com/liferay/portal/bootstrap/ModuleFrameworkImpl.java#L986) can't return the real dir.

**Comment by yuhai** [comment]
ci:test:relevant

**Comment by dantewang** [comment]
@yuhai @ericyanLr Sorry I got lost again here. Could you clarify what is the expected behavior dealing with a symbol link? When to resolve it to the actual file, and when to keep it? It seems we're mixing these.

**Comment by zhangyan0701** [comment]
Integration test [FileInstallConfigTest](https://testray.liferay.com/home/-/testray/case_results/1986130949) was stuck at test testFindConfigurationWithFelixFileNameWhenUsingSymlinkAfterReloadingConfiguration:
```
     [exec] com.liferay.portal.file.install.deploy.test.FileInstallConfigTest > testFindConfigurationWithFelixFileNameWhenUsingSymlinkAfterReloadingConfiguration STARTED
[beanshell] The top level build is alive.
```
Full console log:
https://testray.liferay.com/reports/production/logs/2022-09/test-1-31/test-portal-acceptance-pullrequest(master)/5130/modules-integration-mysql57-jdk8/0/2/jenkins-console.txt.gz?authuser=0

**Comment by ericyanLr** [comment]
Hi @dantewang ,

**Regarding:**
> @yuhai @ericyanLr Sorry I got lost again here. Could you clarify what is the expected behavior dealing with a symbol link? When to resolve it to the actual file, and when to keep it? It seems we're mixing these.

I agree, it is quite confusing. @yuhai might actually have more info at this point, with the testing done on both Linux and Windows.

**Context**
The main issue right now is that there is inconsistency in how the FELIX_FILE_INSTALL_FILENAME is being retrieved/stored for a configuration.

This inconsistency can be observed when encountering one or both of these two factors:
- MODULE_FRAMEWORK_CONFIGS_DIR is symlinked ([LPS-113749](https://issues.liferay.com/browse/LPS-113749))
- Config file is symlinked ([LPS-162706](https://issues.liferay.com/browse/LPS-162706))

**Goal**
The goal is to simply make sure we are using the same value when we are retrieving/storing the FELIX_FILE_INSTALL_FILENAME for each configuration.

I'll put additional details below in case it helps. 

But, I think @yuhai would be able to give you a more accurate answer.

Feel free to let me know if you have any further questions.
Thanks!

----

**Proposed Fix - Info**
My proposed fix was to make changes to the logic in ConfigurationPersistenceManager.java. From my testing, this is where the inconsistency occurs. 

**Analysis**
From my understanding, when a **.config** file is processed, we do the following:
(Ex1 and Ex2 are two different example scenarios)
- Retrieve the canonicalPath of the directory
  - Ex1: [ModuleFrameworkImpl.java#L986](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/core/portal-bootstrap/src/main/java/com/liferay/portal/bootstrap/ModuleFrameworkImpl.java#L986)
  - Ex2: [DirectoryWatcher.java#L89](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/DirectoryWatcher.java#L89) -> [DirectoryWatcher.java#L131-L133](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/DirectoryWatcher.java#L131-L133) -> [Scanner.java#L54](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/Scanner.java#L54) -> [Scanner.java#L147](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/Scanner.java#L147)
- Retrieve file of .config file using the directory w/ canonicalPath
  - Ex1: [ModuleFrameworkImpl.java#L1130](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/core/portal-bootstrap/src/main/java/com/liferay/portal/bootstrap/ModuleFrameworkImpl.java#L1130)
  - Ex2: [DirectoryWatcher.java#L218](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/DirectoryWatcher.java#L218) -> [Scanner.java#L92](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/Scanner.java#L92) -> [Scanner.java#L243](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/Scanner.java#L243)
- Retrieve the absolutePath of the .config file to fetch for existing configuration
  - Ex1: [ConfigurationFileInstaller.java#L92](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/configuration/ConfigurationFileInstaller.java#L92) -> [ConfigurationFileInstaller.java#L310-L316](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/configuration/ConfigurationFileInstaller.java#L310-L316)
  - Ex2: [DirectoryWatcher.java#L221](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/DirectoryWatcher.java#L221) -> [DirectoryWatcher.java#L1008](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/DirectoryWatcher.java#L1008) -> [DirectoryWatcher.java#L649-L670](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/DirectoryWatcher.java#L649-L670) -> [ConfigurationFileInstaller.java#L92](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/configuration/ConfigurationFileInstaller.java#L92) -> [ConfigurationFileInstaller.java#L310-L316](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/configuration/ConfigurationFileInstaller.java#L310-L316)
- Find and retrieve existing configuration using the retrieved path URI string
  - e.g. [ConfigurationFileInstaller.java#L272](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/configuration/ConfigurationFileInstaller.java#L272) -> [ConfigurationFileInstaller.java#L254-L259](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/configuration/ConfigurationFileInstaller.java#L254-L259)
- Set FELIX_FILE_INSTALL_FILENAME configuration (crafted using .config file path)
  - e.g. [ConfigurationFileInstaller.java#L139-L149](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/configuration/ConfigurationFileInstaller.java#L139-L149)

However, when we attempt to **retrieve existing configurations from the database**, we notice a discrepancy: 
- Retrieve configuration from database and it's FELIX_FILE_INSTALL_FILENAME (filename only)
  - e.g. [ConfigurationPersistenceManager.java#L568-L573](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-configuration/portal-configuration-persistence-impl/src/main/java/com/liferay/portal/configuration/persistence/internal/ConfigurationPersistenceManager.java#L568-L573)
- Retrieve the canonical path of the config file associated to the retrieved **(filename)**
  - e.g. [ConfigurationPersistenceManager.java#L614](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-configuration/portal-configuration-persistence-impl/src/main/java/com/liferay/portal/configuration/persistence/internal/ConfigurationPersistenceManager.java#L614) -> [ConfigurationPersistenceManager.java#L410-L413](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-configuration/portal-configuration-persistence-impl/src/main/java/com/liferay/portal/configuration/persistence/internal/ConfigurationPersistenceManager.java#L410-L413)
  - **Notice the difference here**. (the previous scenario retrieved the canonicalPath for just the directory, whereas here, we are retrieving the canonical path of the config file)
- Set FELIX_FILE_INSTALL_FILENAME configuration (crafted using retrieved data from database)
  - e.g. [ConfigurationPersistenceManager.java#L616-L620](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-configuration/portal-configuration-persistence-impl/src/main/java/com/liferay/portal/configuration/persistence/internal/ConfigurationPersistenceManager.java#L616-L620)

And because the FELIX_FILE_INSTALL_FILENAME that is set when retrieving from the database is different, we fail to retrieve the existing configuration when we try to use FELIX_FILE_INSTALL_FILENAME to retrieve the existing configuration, in the scenarios mentioned above

> - Retrieve the absolutePath of the .config file to fetch for existing configuration
>   - Ex1: [ConfigurationFileInstaller.java#L92](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/configuration/ConfigurationFileInstaller.java#L92) -> [ConfigurationFileInstaller.java#L310-L316](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/configuration/ConfigurationFileInstaller.java#L310-L316)
>   - Ex2: [DirectoryWatcher.java#L221](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/DirectoryWatcher.java#L221) -> [DirectoryWatcher.java#L1008](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/DirectoryWatcher.java#L1008) -> [DirectoryWatcher.java#L649-L670](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/DirectoryWatcher.java#L649-L670) -> [ConfigurationFileInstaller.java#L92](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/configuration/ConfigurationFileInstaller.java#L92) -> [ConfigurationFileInstaller.java#L310-L316](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/configuration/ConfigurationFileInstaller.java#L310-L316)
> - Find and retrieve existing configuration using the retrieved path URI string
>   - e.g. [ConfigurationFileInstaller.java#L272](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/configuration/ConfigurationFileInstaller.java#L272) -> [ConfigurationFileInstaller.java#L254-L259](https://github.com/liferay/liferay-portal/blob/7.4.3.43-ga43/modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/configuration/ConfigurationFileInstaller.java#L254-L259)

**Comment by yuhai** [comment]
@dantewang, @ericyanLr, the test was stuck might be caused by my previous "SF, simplify"(use two same configurationPid in the two test methods). I just push force the changes

**Comment by yuhai** [comment]
ci:test:relevant

**Comment by dantewang** [comment]
@yuhai For the issue that `getCanonicalFile()` doesn't resolve symbol link on Windows, see https://bugs.openjdk.org/browse/JDK-8003887. So the usage of `Path.getRealPath()` API in @ericyanLr 's original fix should be adopted to similar places.

**Comment by dantewang** [comment]
Hi @ericyanLr, Hai talked with me and explained the mechanism. So it seems the common usage is to have a real file with any kind of name, and a symbol link in our config dir with configuration pid as the name. If this is the case, then I agree that we should always honor the "file" name in the config dir, whether it's a real file or a link.

On the other hand, based on https://github.com/liferay-core-infra/liferay-portal/pull/1772#issuecomment-1255772932, I asked Hai to keep your fix, and apply the same pattern in similar places. Hai will also try again to run the tests on Windows, because previously he ran `gradlew` command in admin console, while he should run portal in admin console.

cc @tinatian

**Comment by dantewang** [comment]
@yuhai I'm going to close this PR, please resend, thanks!

**Pull request closed without merging**